File: ishtime.rb
#!/usr/bin/env ruby

time = Time.new()

minute_offset = ((time.min - 3) % 60) / 5

said_minutes = [' five past', ' ten past', ' quarter past', ' twenty past', ' twenty-five past', ' half past', ' twenty-five to', ' twenty to', ' quarter to', ' ten to', ' five to']
prefix = said_minutes[minute_offset]

hour = time.hour
hour += 1 if time.min >= 33

said_hours = [' twelve', ' one', ' two', ' three', ' four', ' five', ' six', ' seven', ' eight', ' nine', ' ten', ' eleven']
hour = said_hours[hour % 12]

oclock = ((time.min <= 2 || time.min >= 58) and " o'clock" or '')

puts "It's#{prefix}#{hour}#{oclock}."

File: ishtime.hs
#!/usr/bin/env runhaskell

module Main where

import Data.Time
  ( LocalTime (LocalTime),
    TimeOfDay (TimeOfDay),
    ZonedTime (ZonedTime),
    getZonedTime,
  )
import System.Environment (getArgs)

main :: IO ()
main = do
  time <- parseTime <$> getZonedTime
  args <- unwords <$> getArgs
  putStrLn . ishtime time $
    if args == ""
      then "It's"
      else args

parseTime :: ZonedTime -> (Int, Int)
parseTime (ZonedTime (LocalTime _ (TimeOfDay h m _)) _) = (h, m)

ishtime :: (Int, Int) -> String -> String
ishtime (h, m) msg
  | m <= 2 || m > 57 = concat [msg, " ", hour, " o'clock."]
  | otherwise        = concat [msg, " ", minute, " ", hour, "."]
  where
    minute =
      [ "five past",
        "ten past",
        "a quarter past",
        "twenty past",
        "twenty-five past",
        "half past",
        "twenty-five to",
        "twenty to",
        "a quarter to",
        "ten to",
        "five to"
      ]
        !! ((m - 3) `div` 5)
    hour =
      [ "twelve",
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven"
      ]
        !! (h' `mod` 12)
    h' = h + (m `div` 33) -- Round the hour, if it's "to" something.
